src/services/services.js

```diff
@@ -49,23 +141,169 @@ export const getUserById = async (id) => {
     });
 };
 
-export const getAllMembers = async () => {
-  const docRef = await db.collection('data').doc('users');
+export const getRandomId = () => `${Math.random().toString(36).substr(2, 9)}`;
+
+export const setTaskData = async (data) => {
+  db.collection('data')
+    .doc('tasks')
+    .set(
+      {
+        [data.taskId]: data,
+      },
+      { merge: true },
+    )
+    .then(() => {
+      console.log('Document successfully written!');
+    })
+    .catch((error) => {
+      console.error('Error writing document: ', error);
+    });
+};
+
+export const setUserSubTaskData = async (userId, taskId, data) => {
+  db.collection('data')
+    .doc('memberTasks')
+    .set(
+      {
+        [userId]: { [taskId]: { [data.subTaskId]: data } },
+      },
+      { merge: true },
+    )
+    .then(() => {
+      console.log('Document successfully written!');
+    })
+    .catch((error) => {
+      console.error('Error writing document: ', error);
+    });
+};
+
+export const deleteTaskById = async (taskId) => {
+  const tasks = await db.collection('data').doc('tasks');
+  await tasks.update({
+    [taskId]: firebase.firestore.FieldValue.delete(),
+  });
+};
+
+export const deleteSubTaskById = async (userId, taskId, subTaskId) => {
+  const subTasks = await db.collection('data').doc('memberTasks');
+  const userTasks = await subTasks.get().then((doc) => doc.data()[userId]);
+  const userSubtasks = userTasks[taskId];
+  delete userSubtasks[subTaskId];
+  await subTasks.update({
+    [userId]: userTasks,
+  });
+};
+
+export const getTaskById = async (id) => {
+  const docRef = await db.collection('data').doc('tasks');
   return docRef
     .get()
     .then((doc) => {
       if (doc.exists) {
-        return Object.values(doc.data());
+        return doc.data()[id];
       }
       // doc.data() will be undefined in this case
-      console.log('No such document!');
+      console.log('No such user data, with this ID!');
       return null;
     })
     .catch((error) => {
       console.log('Error getting document:', error);
     });
 };
 
+export const getSubTaskById = async (userId, taskId, subTaskId) => {
+  const docRef = await db.collection('data').doc('memberTasks');
+  return docRef
+    .get()
+    .then((doc) => {
+      if (doc.exists) {
+        return doc.data()[userId][taskId][subTaskId];
+      }
+      // doc.data() will be undefined in this case
+      console.log('No such user data, with this ID!');
+      return null;
+    })
+    .catch((error) => {
+      console.log('Error getting document:', error);
+    });
+};
+
+export const getTasksDataById = async (id) => {
+  const data = [];
+  const docRef = await db.collection('data').doc('tasks');
+  return docRef
+    .get()
+    .then((doc) => {
+      if (doc.exists) {
+        const tasks = Object.values(doc.data());
+        tasks.forEach((item) => {
+          if (item.members[id] === true) {
+            data.push(item);
+          }
+        });
+      }
```

Comment:
use map or reduce

Reply:
I would like to discuss these issues tomorrow, if possible, maybe I do not fully understand the possibilities of map

Reply:
thx, i am use filter here.
```suggestion
 data = tasks.filter((item) => item.members[id] === true);
```